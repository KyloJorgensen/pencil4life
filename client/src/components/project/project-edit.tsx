'use strict';

import * as React from 'react';
import RichTextEditor from 'react-rte';
import {Redirect} from "react-router-dom";
import * as Datetime from 'react-datetime';
import ImageNew from '../image/image-new';
import ImageEdit from '../image/image-edit';
import { IProjectContext, updateProjectParams, projectConsumer } from './project-provider';
import { IUserContext, userConsumer } from '../user/user-provider';
import { Location } from 'history';

export interface ProjectEditProps {
	project: IProjectContext;
	user: IUserContext;
	location: Location;
}

export interface ProjectEditState {
	redirect: boolean;
	required: boolean;
	title: string;
	year: Date;
	imageRequired: boolean;
	coverImage: {
		_imageId: string;
	}; 
	details: any;
	discontinued: boolean;
}

export interface ProjectEditMethods {
	editField: (event: UIEvent) => void;
	handleCheckboxChange: (event: UIEvent) => void;
	yearChanged: (momentDateTime: Date) => void;
	hitKey: (event: UIEvent) => void;
	updateProject: (event: UIEvent) => void;
	updateProjectResult: (error: boolean) => void;
	onRichTextChange: (value) => void;
	redirect: () => void;
	addNewImageResult: (error: boolean, _imageId: string) => void;
}

class ProjectEdit extends React.Component<ProjectEditProps, ProjectEditState> implements ProjectEditMethods {
    constructor(props) {
        super(props);
        this.state = {
        	redirect: false,
        	required: false,
        	title: null,
        	year: null,
        	imageRequired: false,
        	coverImage: {
        		_imageId: null,
        	}, 
        	details: RichTextEditor.createEmptyValue(),
        	discontinued: false,
        };

		this.editField = this.editField.bind(this);
		this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
		this.yearChanged = this.yearChanged.bind(this);
		this.hitKey = this.hitKey.bind(this);
		this.updateProject = this.updateProject.bind(this);
		this.updateProjectResult = this.updateProjectResult.bind(this);
		this.onRichTextChange = this.onRichTextChange.bind(this);
		this.redirect = this.redirect.bind(this);
		this.addNewImageResult = this.addNewImageResult.bind(this);
	}
	
	editField(event) {
		const { name, value } = event.target;
		this.setState((prevState) => {
			prevState[name] = value;
			return prevState;
		});
	}

	handleCheckboxChange(event) {
		const { name, checked } = event.target;
		this.setState((prevState) => {
			prevState[name] = checked;
			return prevState;
		});
	}

	yearChanged(momentDateTime) {
		this.setState((prevState) => {
			return {year: momentDateTime};
		});
	}

	hitKey(event) {
		if (event.key == 'Enter') {
			this.updateProject(event);
		}
	}

	updateProject(event) {
		event.preventDefault();

		const { title, year, details, coverImage, discontinued } = this.state;

		if (!title || !year) {
			this.setState(() => {
				return {required: true};
			});
			return;
		}

		const changes: updateProjectParams = {
			title: title,
			year: year,
			details: details.toString('html'),
			coverImage: coverImage,
			discontinued: discontinued
		};

		const params: updateProjectParams = {_id: this.props.project.project._projectId};
		let vaildKeys = ['title', 'year', 'details', 'coverImage', 'discontinued'];

		vaildKeys.forEach((vaildKey) => {
			if (changes[vaildKey] != this.props[vaildKey]) {
				if (vaildKey == 'coverImage') {
					if (changes.coverImage._imageId !== this.props.project.project.coverImage._imageId) {
						params.coverImage = params.coverImage || {_imageId: null};
						params.coverImage._imageId = changes.coverImage._imageId;
					} 
					return;
				}
				params[vaildKey] = changes[vaildKey];
			}
		});

		this.props.project.updateProject(params, this.updateProjectResult);

		this.setState(() => {
			return {required: false};
		});
	}

	updateProjectResult(error) {
		if (error) {
			this.setState(() => {
				return {required: true};
			});
		} else {
			this.redirect();
		} 
	}

	onRichTextChange(value) {
		this.setState(() => {
			return {details: value};
		});
	}

	redirect() {
		this.setState(() => {
			return {redirect: true};
		});
	}

	addNewImageResult(error, _imageId) {
		if (error) {
			this.setState(() => {
				return {imageRequired: true};
			});
		} else {
			this.setState(() => {
				return {
					coverImage: {
						_imageId: _imageId,
					},
					imageRequired: false,
				};
			});
		} 
	}

    componentWillMount() {
		const { title, year, details, coverImage, discontinued } = this.props.project.project;
		const { _imageId } = coverImage;

		this.setState(() => {
			return {
				title: title,
				year: year,
				details: RichTextEditor.createValueFromString(details, 'html'),
				coverImage: {
					_imageId: _imageId,
				},
				discontinued: discontinued,
			}
		});
    }

	componentDidUpdate(prevProps: ProjectEditProps, prevState) {
		let propsToCheck = ['title', 'year', 'coverImage', 'details', 'discontinued'];
		let _state = null;
		propsToCheck.forEach((propToCheck) => {
			if (prevProps.project.project[propToCheck] !== this.props.project.project[propToCheck]) {
				_state = _state == null ? {} : _state;
				if (propToCheck == 'coverImage') {
					if (prevProps.project.project.coverImage._imageId !== this.props.project.project.coverImage._imageId) {
						_state.coverImage = _state.coverImage || {};
						_state.coverImage._imageId = this.props.project.project.coverImage._imageId;
					}
				} else if (propToCheck == 'details') {
					_state[propToCheck] = RichTextEditor.createValueFromString(this.props.project.project[propToCheck], 'html');
				} else {
					_state[propToCheck] = this.props.project.project[propToCheck];
				}
			}
		})
		if (_state) {
			this.setState(_state);
		}
	}

    componentDidMount() {
    	this.props.project.getProject(this.props.project.project._projectId);
    }

	render() {
		const { addNewImageResult, hitKey, handleCheckboxChange, editField, updateProject, yearChanged, onRichTextChange} = this;
		const { title, coverImage, imageRequired, redirect, required, year, details, discontinued } = this.state;
		const { _imageId } = coverImage;
		const { admin } = this.props.user;
		const { pathname } = this.props.location;


		if (redirect) {
			return (<Redirect to={'/project/item/'+this.props.project.project._projectId} />);
		}

		if (!admin) {
			return (<Redirect to={{pathname: '/login', state: {redirectPath: pathname}}}/>)
		}

		return (
			<div className="project-edit-wrapper" >
				<h3>EDIT PROJECT</h3>
				<label>Cover Image</label>
				{_imageId ? <ImageEdit _imageId={_imageId} updateRedirect={false} /> : <ImageNew required={imageRequired} addNewImageResult={addNewImageResult} />}
				<label>Title{required ? (<span className="errortext" >*</span>) : ''}</label>
				<br/>
				<input type='text' onKeyPress={hitKey} placeholder="Great Project" onChange={editField} name='title' value={title} />
				<br/>
				<label>Year{required ? (<span className="errortext" >*</span>) : ''}</label>
				<br/>
				<Datetime dateFormat="YYYY" timeFormat={false} value={year} onChange={yearChanged} />
				<br/>
				<label>Details</label>
				<br/>
				<RichTextEditor value={details} onChange={onRichTextChange} />
				<label>Discontinued</label>
				<br/>
				<input type='checkbox' checked={!!discontinued} name="discontinued" onChange={handleCheckboxChange} />
				<br/>
				<input type='submit' onClick={updateProject} value='SAVE' />
			</div>
		);			
	}
};

export default userConsumer(projectConsumer(ProjectEdit));