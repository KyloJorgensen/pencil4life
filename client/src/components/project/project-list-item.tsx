'use strict';

import * as React from 'react';
import { Link } from 'react-router-dom';
import * as moment from 'moment';
import * as objectPath from 'object-path';
import ImageTag from '../image/image-tag';
import { projectConsumer, IProjectContext } from './project-provider';

export interface ProjectListItemProps {
	project: IProjectContext;
	_projectId: string;
}

class ProjectListItem extends React.Component<ProjectListItemProps> {
    constructor(props) {
        super(props);
    }

    componentDidMount() {
    	this.props.project.getProject(this.props._projectId);
    }

	render() {

		const { _projectId, title, coverImage } = this.props.project.project;

		let year = moment(this.props.project.project.year).format("YYYY");
		const coverImageElem = objectPath.get(coverImage, '_imageId', false) ? (<ImageTag _imageId={this.props.project.project.coverImage._imageId} />) : '';

		return (
			<div className="project-list-item-wrapper" >				
				<Link className="project-link" to={'/project/item/'+_projectId} >
					<div className="image-wrapper">
						{coverImageElem}
					</div>
					<div className="title-year-wrapper">
						<p className="title">{title}</p>
						<p className="year">{year}</p>
					</div>
				</Link>
			</div>
		);		
	}
};

export default projectConsumer(ProjectListItem);