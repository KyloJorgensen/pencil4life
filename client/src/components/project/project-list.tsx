'use strict';

import * as React from 'react';
import { Link } from 'react-router-dom';
import ProjectListItem from './project-list-item';
import { IProjectContext, projectConsumer, getProjectsQuery } from './project-provider';
import { IUserContext, userConsumer } from '../user/user-provider';


export interface ProjectListProps {
	project: IProjectContext;
	user: IUserContext;
	limit: number;
	discontinued: boolean;
	displayDetails?: boolean;
}

class ProjectList extends React.Component<ProjectListProps> {
	public static defaultProps = {
		displayDetails: true,
	};
    constructor(props) {
        super(props);
    }

    componentDidMount() {
    	let query: getProjectsQuery = {};
    	if (this.props.limit) {
    		query.limit = this.props.limit;
    	}
    	this.props.project.getProjects(query);
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
    	let update = false;
    	let prevDiscontinued = prevProps.discontinued || false;
    	let currentDiscontinued = this.props.discontinued || false;
    	if (prevDiscontinued != currentDiscontinued) {
	    	update = true;
    	}

    	if (update) {
    		let query = {
    			discontinued: this.props.discontinued || false,
    			limit: this.props.limit || 10,
    		}
    		this.props.project.getProjects(query);
    	}

    }

	render() {
		const { displayDetails } = this.props;
		let ProjectList = [];
		if (this.props.project.projects) {
            this.props.project.projects.forEach((projectKey) => {
				ProjectList.push(<ProjectListItem key={projectKey} _projectId={projectKey} displayDetails={displayDetails} />)
			});
		}
		return (
			<div className="project-wrapper" >
				{this.props.user.admin ? (
					<div>
						<Link to="/project/new">NEW PROJECT</Link>
						<br/>
					</div>
				) : ''}
				<ul className='project-list' >
					{ProjectList}
				</ul>
			</div>
		);			
	}
};

export default userConsumer(projectConsumer(ProjectList));