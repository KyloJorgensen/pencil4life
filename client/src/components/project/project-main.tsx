'use strict';

import * as React from 'react';
import ProjectList from './project-list';
import { userConsumer, IUserContext } from '../user/user-provider';
import { IProjectContext, projectConsumer } from './project-provider';

export interface ProjectMainProps {
    user: IUserContext;
    project: IProjectContext;
}

export interface ProjectMainState {
    discontinued: boolean;
}

export interface ProjectMainMethods {
    handleCheckboxChange: (event: UIEvent) => void;
}

class ProjectMain extends React.Component<ProjectMainProps, ProjectMainState> implements ProjectMainMethods {
    constructor(props) {
        super(props);
        this.state = {
        	discontinued: false,
        };

		this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
    }

    handleCheckboxChange(event) {
        const { name, checked } = event.target;
        this.setState((prevState) => {
            prevState[name] = checked; 
            return prevState;
        });
    }

	render() {
        let adminOptions = this.props.user.admin ? (
            <div>
                <h5>Admin Options</h5>
                <label>Discontinued</label><input type='checkbox' checked={!!this.state.discontinued} name="discontinued" onChange={this.handleCheckboxChange} />
            </div>
        ) : '';

        return (
			<div className="project-wrapper" >
                {adminOptions}
				<ProjectList discontinued={this.state.discontinued} />
			</div>
		);			
	}
};

export default userConsumer(projectConsumer(ProjectMain));