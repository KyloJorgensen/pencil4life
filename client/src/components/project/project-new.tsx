'use strict';

import * as React from 'react';
import RichTextEditor from 'react-rte';
import {Redirect} from 'react-router-dom';
import * as Datetime from 'react-datetime';
import { Location } from 'history';
import ImageEditingTool from '../image/image-editing-tool';
import { IUserContext, userConsumer } from '../user/user-provider';
import { IProjectContext, addProjectParams, projectConsumer } from './project-provider';

export interface ProjectNewProps {
	user: IUserContext;
	project: IProjectContext;
	location: Location;
}

export interface ProjectNewState {
	_projectId: string;
	required: boolean;
	imageRequired: boolean;
	title: string;
	coverImage: {
		_imageId: string;
	};
	year: Date;
	details: any;
}

export interface ProjectNewMethods {
	hitKey: (event: UIEvent) => void;
	yearChanged: (date: Date) => void;
	addNewProject: (event: UIEvent) => void;
	addNewProjectResult: (error: boolean, _projectId: string) => void;
	onRichTextChange: (value: string) => void;
	handleCoverImageChange: (_imageId) => void;
	removeImage: () => void;
}


class ProjectNew extends React.Component<ProjectNewProps, ProjectNewState> implements ProjectNewMethods {
    constructor(props) {
        super(props);
        this.state = {
        	_projectId: null,
        	required: false,
        	imageRequired: false,
        	title: null,
        	coverImage: {
        		_imageId: null,
        	}, 
        	year: new Date(),
        	details: RichTextEditor.createEmptyValue(),
		};
		

		this.hitKey = this.hitKey.bind(this);
		this.yearChanged = this.yearChanged.bind(this);
		this.addNewProject = this.addNewProject.bind(this);
		this.addNewProjectResult = this.addNewProjectResult.bind(this);
		this.onRichTextChange = this.onRichTextChange.bind(this);
		this.handleCoverImageChange = this.handleCoverImageChange.bind(this);
		this.removeImage = this.removeImage.bind(this);
	}

	hitKey(event) {
		if (event.key == 'Enter') {
			this.addNewProject(event);
		}
	}

	yearChanged(date) {
		this.setState(() => {
			return {year: date};
		});
	}

	titleRef: React.RefObject<HTMLInputElement> = React.createRef();

	addNewProject(event) {
		event.preventDefault();
		const title = this.titleRef.current.value;
		const { year, details, coverImage } = this.state;
		if (!title || !year) {
			this.setState(() => {
				return {required: true};
			});
			return;
		}

		const newProject: addProjectParams = {
			title: title,
			year: year,
			details: details.toString('html'),
		}

		if (coverImage._imageId != null) {
			newProject.coverImage = newProject.coverImage || {_imageId: ''};
			newProject.coverImage._imageId = coverImage._imageId;
		}

		this.props.project.addProject(newProject, this.addNewProjectResult);

		this.setState(() => {
			return {required: false};
		});
	}

	addNewProjectResult = (error, _projectId) => {
		if (error) {
			this.setState(() => {
				return {required: true};
			});
		} else {
			this.setState(() => {
				return {_projectId: _projectId};
			});
		} 
	}

	onRichTextChange = (value) => {
		this.setState(() => {
			return {details: value};
		});
	}

	handleCoverImageChange(_imageId) {
		this.setState(() => {
			return {
				coverImage: {
					_imageId: _imageId,
				},
			};
		});
	}

	removeImage() {
		this.setState(() => {
			return {
				coverImage: {
					_imageId: null,
				},
			}
		});
	}
	
	render() {
		const { handleCoverImageChange, removeImage, addNewProject, hitKey, yearChanged, onRichTextChange } = this;
		const { coverImage,  _projectId, required, details, year } = this.state;
		const { _imageId } = coverImage;
		const { admin } = this.props.user;
		const { pathname } = this.props.location;

		if (_projectId) {
			return (<Redirect to={'/project/item/'+this.state._projectId} />);
		}
		if (!admin) {
			return (<Redirect to={{pathname: '/login', state: {redirectPath: pathname}}}/>)
		}
		console.log(_imageId);
		return (
			<div className="project-new-wrapper" >
				<h3>New Project</h3>
				<label>Cover Image</label>
				<ImageEditingTool _imageId={_imageId} onChange={handleCoverImageChange} removeImage={removeImage} />				
				<label>Title{required ? (<span className="errortext" >*</span>) : ''}</label>
				<br/>
				<input type='text' onKeyPress={hitKey} placeholder="Great Project" ref={this.titleRef} />
				<br/>
				<label>Year{required ? (<span className="errortext" >*</span>) : ''}</label>
				<br/>
				<Datetime dateFormat="YYYY" timeFormat={false} onChange={yearChanged} value={year} />
				<br/>
				<label>Details</label>
				<br/>
				<RichTextEditor value={details} onChange={onRichTextChange} />
				<br/>
				<input type='submit' onClick={addNewProject} value='SAVE' />
			</div>
		);			
	}
};

export default userConsumer(projectConsumer(ProjectNew));