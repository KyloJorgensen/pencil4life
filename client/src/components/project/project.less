.project-list-item-wrapper {

}

.project-wrapper {
	.project-list {	
		display: grid;
	    grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
		grid-template-rows: minmax(100px, 1fr);
		grid-gap: 2rem;
	    .project-link {    
			display: grid;
			grid-template-rows: 1fr auto;
			width: 100%;
			height: 100%;
			color: black;
			border-color: #ffffff;
			border-style: solid;
	    }
	    .project-link:hover {
			border-style: solid;
			border-color: black;
			transition: border-color 2s linear;
		}
		.title-year-wrapper {
			align-self: end;
			> .title, > .year {
			    margin: 0.25em;
			    display: inline-block;
			}
			> .year {
				float:right;
			}
		}
	}
}

