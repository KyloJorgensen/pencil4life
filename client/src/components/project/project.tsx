'use strict';

import * as React from 'react';
import RichTextEditor from 'react-rte';
import { Link, Switch, Route, NavLink } from "react-router-dom";
import * as moment from 'moment';
import { projectConsumer, IProjectContext } from './project-provider';
import { IUserContext, userConsumer } from '../user/user-provider';
import ImageTag from '../image/image-tag';
import ProjectPageList from './page/project-page-list';
import { projectPageProvider } from './page/project-page-provider';
import ProjectPage from './page/project-page';
import ProjectPageNew from './page/project-page-new';
import ProjectPageEdit from './page/project-page-edit';
import styled from 'styled-components';

const Wrapper = styled.div`
	display: grid;
	grid-row-gap: 1em;
	> .title-year-wrapper {
		display: flex;
		justify-content: space-between;
	}
	> img {
		max-width: 100%;
		margin: 0 auto;
	}

`;

export interface ProjectProps {
	project: IProjectContext;
	user: IUserContext;
}

export interface ProjectState {
	discontinued: boolean;
}

class Project extends React.Component<ProjectProps, ProjectState> {
    constructor(props) {
        super(props);
        this.state = {
        	discontinued: false,
        };

		this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
    }

    handleCheckboxChange(event) {
        const { name, checked } = event.target;
        this.setState((prevState) => {
            prevState[name] = checked; 
            return prevState;
        });
    }

    componentDidMount() {
    	this.props.project.getProject(this.props.project.project._projectId);
    }

	render() {
		const { _projectId, title, year, details, coverImage } = this.props.project.project;
		const { _imageId } = coverImage;
		const { admin } = this.props.user;
		const { discontinued } = this.state;

		const formatedYear = moment(year).format("YYYY");
		const coverImageElem = _imageId? (<ImageTag _imageId={_imageId} />) : '';
		const convertedDetails = RichTextEditor.createValueFromString(details, 'html');
		return (
			<Wrapper className="project-wrapper" >
				{admin ? <p className="text-right" ><Link to={'/project/edit/'+_projectId}>EDIT</Link></p> : ''}
				<div className="title-year-wrapper">
					<NavLink to={'/project/item/'+_projectId} ><h2 className="title">{title}</h2></NavLink>
					<p className="year">{formatedYear}</p>
				</div>
				{coverImageElem}
				{ details != '<p><br></p>' ? <RichTextEditor value={convertedDetails} readOnly={true} /> : ''}
				{admin ? (
					<div>
						<Link to={"/project/new/"+_projectId}>NEW PAGE</Link>
						<label>Discontinued</label><input type='checkbox' checked={!!this.state.discontinued} name="discontinued" onChange={this.handleCheckboxChange} />
					</div>
				) : ''}
				<ProjectPageList discontinued={discontinued} />
				<Switch>
					<Route exact path="/project/new/:_projectId/" component={ProjectPageNew} />
					<Route exact path="/project/item/:_projectId/:_projectPageId" component={ProjectPage} />
					<Route exact path="/project/edit/:_projectId/:_projectPageId" component={ProjectPageEdit} />				
				</Switch>
			</Wrapper>
		);			
	}
};


export { Project };

export default userConsumer(projectConsumer(projectPageProvider(Project)));