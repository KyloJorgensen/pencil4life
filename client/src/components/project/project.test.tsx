'use strict';

import * as React from 'react';
import { shallow } from 'enzyme';
import * as RichTextEditor from 'react-rte'
import { Project, ProjectProps } from './project';

const _defaultProps: ProjectProps = {
    project: {    
        getProjects: (params) => {},
        getProject: (params) => {},
        addProject: (params) => {},
        updateProject: (params, callback) => {},
        projects: [],
        projectList: {},
        page: 0,
        totalPages: 0,
        limit: 0,
        total: 0,
        pageTotal: 0,
        project: {
            _projectId: '',
            title: '',
            year: new Date(0),
            details: RichTextEditor.createEmptyValue().toString('html'),
            discontinued: false,
            coverImage: {
                _imageId: '',
            },
        }
    },
    user: {
        signup: () => {},
          getUser: () => {},
          login: () => {},
          logout: () => {},
          updateUser: () => {},
          updatePassword: () => {},
          email: null,
          firstname: null,
          lastname: null,
          username: null,
          userAccess: false,
          admin: false,
    }
};

describe('Render Project', () => {
	it('matches the snapshot', () => {
		let wrapper = shallow(<Project {..._defaultProps} />);
		expect(wrapper).toMatchSnapshot();
	});
});