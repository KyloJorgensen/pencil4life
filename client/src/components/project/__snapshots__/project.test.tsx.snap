// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Project matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Project
    project={
      Object {
        "addProject": [Function],
        "getProject": [Function],
        "getProjects": [Function],
        "limit": 0,
        "page": 0,
        "pageTotal": 0,
        "project": Object {
          "_projectId": "",
          "coverImage": Object {
            "_imageId": "",
          },
          "details": "<p><br></p>",
          "discontinued": false,
          "title": "",
          "year": 1970-01-01T00:00:00.000Z,
        },
        "projectList": Object {},
        "projects": Array [],
        "total": 0,
        "totalPages": 0,
        "updateProject": [Function],
      }
    }
    user={
      Object {
        "admin": false,
        "email": null,
        "firstname": null,
        "getUser": [Function],
        "lastname": null,
        "login": [Function],
        "logout": [Function],
        "signup": [Function],
        "updatePassword": [Function],
        "updateUser": [Function],
        "userAccess": false,
        "username": null,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        "",
        <div
          className="title-year-wrapper"
        >
          <NavLink
            activeClassName="active"
            aria-current="page"
            to="/project/item/"
          >
            <h2
              className="title"
            >
              
            </h2>
          </NavLink>
          <p
            className="year"
          >
            1969
          </p>
        </div>,
        "",
        "",
        <UserConsumer />,
        <Switch>
          <Route
            component={[Function]}
            exact={true}
            path="/project/new/:_projectId/"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/project/item/:_projectId/:_projectPageId"
          />
          <Route
            component={[Function]}
            exact={true}
            path="/project/edit/:_projectId/:_projectPageId"
          />
        </Switch>,
      ],
      "className": "project-wrapper",
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/project/item/"
            >
              <h2
                className="title"
              >
                
              </h2>
            </NavLink>,
            <p
              className="year"
            >
              1969
            </p>,
          ],
          "className": "title-year-wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeClassName": "active",
              "aria-current": "page",
              "children": <h2
                className="title"
              >
                
              </h2>,
              "to": "/project/item/",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "className": "title",
              },
              "ref": null,
              "rendered": "",
              "type": "h2",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "1969",
              "className": "year",
            },
            "ref": null,
            "rendered": "1969",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "",
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              component={[Function]}
              exact={true}
              path="/project/new/:_projectId/"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/project/item/:_projectId/:_projectPageId"
            />,
            <Route
              component={[Function]}
              exact={true}
              path="/project/edit/:_projectId/:_projectPageId"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/project/new/:_projectId/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/project/item/:_projectId/:_projectPageId",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
              "exact": true,
              "path": "/project/edit/:_projectId/:_projectPageId",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          "",
          <div
            className="title-year-wrapper"
          >
            <NavLink
              activeClassName="active"
              aria-current="page"
              to="/project/item/"
            >
              <h2
                className="title"
              >
                
              </h2>
            </NavLink>
            <p
              className="year"
            >
              1969
            </p>
          </div>,
          "",
          "",
          <UserConsumer />,
          <Switch>
            <Route
              component={[Function]}
              exact={true}
              path="/project/new/:_projectId/"
            />
            <Route
              component={[Function]}
              exact={true}
              path="/project/item/:_projectId/:_projectPageId"
            />
            <Route
              component={[Function]}
              exact={true}
              path="/project/edit/:_projectId/:_projectPageId"
            />
          </Switch>,
        ],
        "className": "project-wrapper",
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <NavLink
                activeClassName="active"
                aria-current="page"
                to="/project/item/"
              >
                <h2
                  className="title"
                >
                  
                </h2>
              </NavLink>,
              <p
                className="year"
              >
                1969
              </p>,
            ],
            "className": "title-year-wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "active",
                "aria-current": "page",
                "children": <h2
                  className="title"
                >
                  
                </h2>,
                "to": "/project/item/",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "className": "title",
                },
                "ref": null,
                "rendered": "",
                "type": "h2",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "1969",
                "className": "year",
              },
              "ref": null,
              "rendered": "1969",
              "type": "p",
            },
          ],
          "type": "div",
        },
        "",
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                component={[Function]}
                exact={true}
                path="/project/new/:_projectId/"
              />,
              <Route
                component={[Function]}
                exact={true}
                path="/project/item/:_projectId/:_projectPageId"
              />,
              <Route
                component={[Function]}
                exact={true}
                path="/project/edit/:_projectId/:_projectPageId"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "exact": true,
                "path": "/project/new/:_projectId/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "exact": true,
                "path": "/project/item/:_projectId/:_projectPageId",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
                "exact": true,
                "path": "/project/edit/:_projectId/:_projectPageId",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
